### PR TITLE
PHPStan add support of magic methods of Data Object.

### LINKED ISSUE #28303: [Issue] PHPStan add support of magic methods of Data Object.
This issue is automatically created based on existing pull request: magento/magento2#27905: PHPStan add support of magic methods of Data Object.

**Preconditions:**
---------
- Add extension to support of 'get/set/has/uns' 'magic' methods (with usage of __call) for DataObject;
- Add support for SessionManager (it forwards all calls to DataObject container);
- Add test coverage for extensions;
- fix outdated tests for Filtered Error check;
- increase PHPStan check level 0 -> 1.

**Steps to reproduce:**
N/a

**Actual Result:**

**Expected Result:**
Q:
- Are tests from **dev/tests/static/framework/tests** folder also included to test builds?

### PR DESCRIPTION
**Pull request by swnsma**
- Add extension to support of 'get/set/has/uns' 'magic' methods (with usage of __call) for DataObject;
- Add support for SessionManager (it forwards all calls to DataObject container);
- Add test coverage for extensions;
- fix outdated tests for Filtered Error check;
- increase PHPStan check level 0 -> 1.

Q:
- Are tests from **dev/tests/static/framework/tests** folder also included to test builds?

### Resolved issues:
1. [x] resolves magento/magento2#28303: PHPStan add support of magic methods of Data Object.

### CODE REVIEW

**Comment by viktym** [comment]
Hi @swnsma 
Thank you for your contribution
Increasing the PHPStan level from 0 to 1 involves activation next additional rules and checks:
```
parameters:
	checkMaybeUndefinedVariables: true
	checkExtraArguments: true
	reportMagicMethods: true
	reportMagicProperties: true


conditionalTags:
	PHPStan\Rules\Variables\VariableCertaintyNullCoalesceRule:
		phpstan.rules.rule: %featureToggles.nullCoalesce%

rules:
	- PHPStan\Rules\Classes\UnusedConstructorParametersRule
	- PHPStan\Rules\Constants\ConstantRule
	- PHPStan\Rules\Functions\UnusedClosureUsesRule
	- PHPStan\Rules\Variables\VariableCertaintyInIssetRule

services:
	-
		class: PHPStan\Rules\Variables\VariableCertaintyNullCoalesceRule
```

Can you run PHPStan check using level 1 on CE/EE/B2B Magento codebase and compare results with level 0 (create a diff) ?
Please, make sure that all reported errors can be fixed when Magento developers will face them in their PRs. Or, possibly, some files should be added in blacklist at once.
Also, please check if new PHPStan checks intersect with PHPMD and PHPCS checks. We should avoid double suppression (like unused method argument etc.).

**Review by viktym — COMMENTED** [review]
Hi @swnsma 
Thank you for your contribution
Increasing the PHPStan level from 0 to 1 involves activation next additional rules and checks:
```
parameters:
	checkMaybeUndefinedVariables: true
	checkExtraArguments: true
	reportMagicMethods: true
	reportMagicProperties: true


conditionalTags:
	PHPStan\Rules\Variables\VariableCertaintyNullCoalesceRule:
		phpstan.rules.rule: %featureToggles.nullCoalesce%

rules:
	- PHPStan\Rules\Classes\UnusedConstructorParametersRule
	- PHPStan\Rules\Constants\ConstantRule
	- PHPStan\Rules\Functions\UnusedClosureUsesRule
	- PHPStan\Rules\Variables\VariableCertaintyInIssetRule

services:
	-
		class: PHPStan\Rules\Variables\VariableCertaintyNullCoalesceRule
```

Can you run PHPStan check using level 1 on CE/EE/B2B Magento codebase and compare results with level 0 (create a diff) ?
Please, make sure that all reported errors can be fixed when Magento developers will face them in their PRs. Or, possibly, some files should be added in blacklist at once.
Also, please check if new PHPStan checks intersect with PHPMD and PHPCS checks. We should avoid double suppression (like unused method argument etc.).

**Comment by swnsma** [comment]
Hi, @viktym!

Please find the diff: https://www.diffchecker.com/rHxwkJVM
In total there are 1604 vs 2356 errors.
I have used CE + MSI + EE + B2B codebase (lib + app/code + dev + setup).
From the diff I can sum up that for Magento relevant next new checks:
- incorrect number of arguments passed;
- a variable might not be defined check;
- **isset** usage;
- access to the undefined property;
- anonymous function and use and closure;
- call to the undefined method;
- find undefined constants;

From all these violations, I see problems with constants (most of the problems are related to tests).
1) Constants which is defined by tests settings:
- TESTS_WEB_API_ADAPTER
- TESTS_BASE_URL
- TESTS_XDEBUG_ENABLED
- TESTS_XDEBUG_SESSION
As they did not declare in PHP directly, I suggest adding these constants to "ignoreErrors" in phpstan.neon
2) Constants defined by tests fixtures:
- STAGING_UPDATE_FIXTURE_ID
- FIXTURE_ATTRIBUTE_USER_DEFINED_CUSTOMER_NAME
- FIXTURE_ATTRIBUTE_USER_DEFINED_CUSTOMER_FRONTEND_LABEL
I suggest to skip them and just use "phpstan:ignore"
3) Constants, defined by tests setup:
- INTEGRATION_TESTS_DIR
- MAGENTO_MODULES_PATH
- TESTS_MODULES_PATH
- TESTS_INSTALLATION_DB_CONFIG_FILE
These constants are defined on the fly and PHPStan needs to have files with constants definition in autoload.
We need to autoload: dev/tests/setup-integration/framework/bootstrap.php and dev/tests/api-functional/framework/bootstrap.php
But I suggest also to add these constants to "ignoreErrors".
4) Constants defined by PHP CS
- T_STRING_CONCAT                  
- T_OPEN_CURLY_BRACKET
- T_CLOSE_CURLY_BRACKET            
- T_OPEN_PARENTHESIS                            
- T_CLOSE_PARENTHESIS                           
- T_STYLE                              
- T_ASPERAND                         
- T_OPEN_CURLY_BRACKET                       
- T_CLOSE_CURLY_BRACKET                    
- T_OPEN_CURLY_BRACKET       
- T_STYLE                    
- T_SEMICOLON
- T_ASPERAND
- T_COLON                              
- T_SEMICOLON 
- T_COLON     
- T_STYLE
All these constants are defined in vendor/squizlabs/php_codesniffer/src/Util/Tokens.php
I am not sure in this point, probably, we can include this file to autoload.
5) Constants from file dev/tools/UpgradeScripts/pre_composer_update_2.3.php
- INFO
- WARN
- ERROR
I suggest skipping this point.

What do you think about this?

**Comment by viktym** [comment]
@swnsma great job!
I'll review the diff and provide feedback till the end of the week

**Comment by viktym** [comment]
Hi @swnsma 
I agree with your suggestions 1-3

Also, I recommend to ignore constants from 4 using pattern: `#Constant T_[A-Z_]+  not found.# `

Number 5 you can add in blacklist:
> dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/blacklist/common.txt

Please proceed with implementation of these recommendations and fix the failed Integrity test as well:

> Passed: 32740, Failed: 1, Incomplete: 0, Skipped: 0.
> Data set: dev/tests/static/framework/tests/unit/testsuite/Magento/PhpStan/Reflection/DataObject/MagicMethodsTest.php
> Location of /var/www/html/dev/tests/static/framework/tests/unit/testsuite/Magento/PhpStan/Reflection/DataObject/MagicMethodsTest.php does not match formal namespace: Magento\PhpStan\Reflection\DataObject 
>

**Comment by swnsma** [comment]
Hi @viktym!
Done!

**Comment by viktym** [comment]
Hi @swnsma 
PR looks good for me
Don't have any objections

File: dev/tests/static/framework/Magento/PhpStan/Reflection/Php/DataObjectClassReflectionExtension.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    public function hasMethod(ClassReflection $classReflection, string $methodName): bool
+    {
+        // Workaround due to annotation extension always loads last.
+        if ($this->annotationsMethodsClassReflectionExtension->hasMethod($classReflection, $methodName)) {
+            // In case when annotation already available for the method, we will not use 'magic methods' approach.
+            return false;
+        }
+
+        if ($classReflection->isSubclassOf(DataObject::class) || $classReflection->getName() == DataObject::class) {
+            return in_array(substr($methodName, 0, 3), self::MAGIC_METHODS_PREFIXES);
```
Comment (lenaorobei) [inline dev/tests/static/framework/Magento/PhpStan/Reflection/Php/DataObjectClassReflectionExtension.php]: It would be nice to get rid of magic numbers.

Comment (swnsma) [reply dev/tests/static/framework/Magento/PhpStan/Reflection/Php/DataObjectClassReflectionExtension.php]: Cleaned.

**Comment by lenaorobei** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
@magento run all tests

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-7566](https://jira.corp.magento.com/browse/ENGCOM-7566) has been created to process this Pull Request
:eight_spoked_asterisk: @lenaorobei, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
QA not applicable

**Comment by slavvka** [comment]
@magento create issue

**Comment by swnsma** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
HI @swnsma, could you please look through a failed static tests?

**Comment by swnsma** [comment]
Fixed.

BTW.
I think this PR, at least, deserves 'Award: test coverage' label, as it introduce tests coverage for the new functionality.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Pull request merged**